### PR TITLE
Allow to use TIMESTAMP as output column type for Kafka message timestamp

### PR DESCRIPTION
**Pull request by millin**
Adding the ability to use TIMESTAMP as the output column type for any field containing milliseconds Unix epoch timestamp (incl. message timestamp).

It was previously impossible to use TIMESTAMP as output column type
```sql
CREATE SCHEMA RND;

CREATE OR REPLACE TABLE RND.TEST_KAFKA_TIMESTAMP (
  DTTM TIMESTAMP,
  JSON VARCHAR(2000000),
  KAFKA_PARTITION DECIMAL(18, 0),
  KAFKA_OFFSET    DECIMAL(36, 0)
);

IMPORT INTO RND.TEST_KAFKA_TIMESTAMP (DTTM, JSON, KAFKA_PARTITION, KAFKA_OFFSET)
FROM SCRIPT KAFKA_EXTENSION.KAFKA_CONSUMER
WITH
 BOOTSTRAP_SERVERS = 'kafka-01:9092'
 TOPIC_NAME = 'test_topic'
 RECORD_FIELDS = 'timestamp,value'
 TABLE_NAME = 'RND.TEST_KAFKA_TIMESTAMP'
 RECORD_VALUE_FORMAT = 'string'
 GROUP_ID = 'exasol-kafka-test'
;
```
results in an error
```
[2021-08-25 16:21:15] [22002] VM error: F-UDF-CL-LIB-1127: F-UDF-CL-SL-JAVA-1002: F-UDF-CL-SL-JAVA-1013:
[2021-08-25 16:21:15] com.exasol.ExaUDFException: F-UDF-CL-SL-JAVA-1080: Exception during run
[2021-08-25 16:21:15] com.exasol.cloudetl.kafka.KafkaConnectorException: Error consuming Kafka topic 'test_topic' data. It occurs for partition '1' in node '0' and vm '139980029142784' Cause: E-UDF-CL-SL-JAVA-1109: emit column 'DTTM' is of type TIMESTAMP but data given have type java.lang.Long
[2021-08-25 16:21:15] com.exasol.cloudetl.kafka.consumer.KafkaRecordConsumer.emit(KafkaRecordConsumer.scala:64)
[2021-08-25 16:21:15] com.exasol.cloudetl.kafka.KafkaTopicDataImporter$.run(KafkaTopicDataImporter.scala:43)
[2021-08-25 16:21:15] com.exasol.cloudetl.kafka.KafkaTopicDataImporter.run(KafkaTopicDataImporter.scala)
[2021-08-25 16:21:15] com.exasol.ExaWrapper.run(ExaWrapper.java:196)
[2021-08-25 16:21:15] Caused by: com.exasol.ExaDataTypeException: E-UDF-CL-SL-JAVA-1109: emit column 'DTTM' is of type TIMESTAMP but data given have type java.lang.Long
[2021-08-25 16:21:15] com.exasol.ExaIteratorImpl.emit(ExaIteratorImpl.java:124)
[2021-08-25 16:21:15] com.exasol.cloudetl.kafka.consumer.KafkaRecordConsumer.$anonfun$emitRecords$1(KafkaRecordConsumer.scala:110)
[2021-08-25 16:21:15] com.exasol.cloudetl.kafka.consumer.KafkaRecordConsumer.$anonfun$emitRecords$1$adapted(KafkaRecordConsumer.scala:101)
[2021-08-25 16:21:15] scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[2021-08-25 16:21:15] scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[2021-08-25 16:21:15] scala.collection.AbstractIterable.foreach(Iterable.scala:919)
[2021-08-25 16:21:15] com.exasol.cloudetl.kafka.consumer.KafkaRecordConsumer.emitRecords(KafkaRecordConsumer.scala:101)
[2021-08-25 16:21:15] com.exasol.cloudetl.kafka.consumer.KafkaRecordConsumer.emit(KafkaRecordConsumer.scala:52)
[2021-08-25 16:21:15] 	... 7 more
```

### CODE REVIEW

**Comment by morazow** [comment]
Hello @millin,

I have refactored and added [an integration test](https://github.com/millin/kafka-connector-extension/blob/convert_timestamp/src/it/scala/com/exasol/cloudetl/kafka/docker/KafkaImportIT.scala#L43-L58) for your changes. Please have a look and feel free to change them (It needed updates to include Avro record producer).

Please also address @baunz suggestions, there are seem to some test failures (from first glance related to empty rows, out of bounds errors). Unfortunately, CI is not run for external pull requests. But you can run `scripts/ci.sh` locally before pushing, it is  same as in CI excluding only Sonar checks.

Thanks for your contributions!

**Comment by morazow** [comment]
Fixes #51

File: src/main/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumer.scala [hunk 1]
```diff
@@ -113,8 +114,9 @@ class KafkaRecordConsumer(
       )
       val columnsCount = tableColumnCount - metadata.size
       val rowValues = RowBuilder.buildRow(recordFieldSpecifications, record, columnsCount)
-      val rows: Seq[Any] = rowValues ++ metadata
-      iterator.emit(rows: _*)
+      val row: Seq[Any] = rowValues ++ metadata
+      val converted_row: Seq[Any] = fieldConverter.convertRow(row)
+      iterator.emit(converted_row: _*)
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumer.scala:119]:
```suggestion
      val convertedRow: Seq[Any] = fieldConverter.convertRow(row)
      iterator.emit(convertedRow: _*)
```

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/FieldConverter.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import java.sql.Timestamp
+
+class FieldConverter(
+  outputColumnTypes: Seq[Class[_]]
+) {
+
+  final def convertRow(row: Seq[Any]): Seq[Any] =
+    row.zipWithIndex.map(x => convert(outputColumnTypes(x._2), x._1))
+
+  final def convert(column_type: Class[_], value: Any): Any =
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/FieldConverter.scala:12]:
```suggestion
  final def convert(columnType: Class[_], value: Any): Any =
```

**Review by morazow — CHANGES_REQUESTED** [review]

**Comment by millin** [comment]
Hello,
I have applied suggestions. I'm working on test errors now

**Comment by millin** [comment]
@morazow I had two options to correct the tests: ignore the absence of meta or mock it. I decided that it's better to mock.
I also corrected new test for timestamp, since it did not pass for me due to different time zones.

Locally I have only one failed test `run throws if it cannot create KafkaConsumer`, 
but looks like it doesn't fail due to my changes:

```verilog
[pool-1-thread-1-ScalaTest-running-KafkaTopicMetadataReaderIT] WARN org.apache.kafka.clients.consumer.ConsumerConfig - The configuration 'schema.registry.url' was supplied but isn't a known config.
[pool-1-thread-1-ScalaTest-running-KafkaTopicMetadataReaderIT] INFO org.apache.kafka.common.utils.AppInfoParser - Kafka version: 6.2.0-ccs
[pool-1-thread-1-ScalaTest-running-KafkaTopicMetadataReaderIT] INFO org.apache.kafka.common.utils.AppInfoParser - Kafka commitId: 1a5755cf9401c84f
[pool-1-thread-1-ScalaTest-running-KafkaTopicMetadataReaderIT] INFO org.apache.kafka.common.utils.AppInfoParser - Kafka startTimeMs: 1630394054721
[pool-1-thread-1-ScalaTest-running-KafkaTopicMetadataReaderIT] WARN org.apache.kafka.clients.NetworkClient - [Consumer clientId=consumer-EXASOL_KAFKA_UDFS_CONSUMERS-5, groupId=EXASOL_KAFKA_UDFS_CONSUMERS] Connection to node -1 (kafka01.internal/127.0.0.1:9092) could not be established. Broker may not be available.
[pool-1-thread-1-ScalaTest-running-KafkaTopicMetadataReaderIT] WARN org.apache.kafka.clients.NetworkClient - [Consumer clientId=consumer-EXASOL_KAFKA_UDFS_CONSUMERS-5, groupId=EXASOL_KAFKA_UDFS_CONSUMERS] Bootstrap broker kafka01.internal:9092 (id: -1 rack: null) disconnected
...
** HERE ARE MANY IDENTICAL LINES AS ABOVE AND BELOW **
...
[pool-1-thread-1-ScalaTest-running-KafkaTopicMetadataReaderIT] WARN org.apache.kafka.clients.NetworkClient - [Consumer clientId=consumer-EXASOL_KAFKA_UDFS_CONSUMERS-5, groupId=EXASOL_KAFKA_UDFS_CONSUMERS] Connection to node -1 (kafka01.internal/127.0.0.1:9092) could not be established. Broker may not be available.
[pool-1-thread-1-ScalaTest-running-KafkaTopicMetadataReaderIT] WARN org.apache.kafka.clients.NetworkClient - [Consumer clientId=consumer-EXASOL_KAFKA_UDFS_CONSUMERS-5, groupId=EXASOL_KAFKA_UDFS_CONSUMERS] Bootstrap broker kafka01.internal:9092 (id: -1 rack: null) disconnected
[info] - run throws if it cannot create KafkaConsumer *** FAILED ***
[info]   Expected exception com.exasol.cloudetl.kafka.KafkaConnectorException to be thrown, but org.apache.kafka.common.errors.TimeoutException was thrown (KafkaTopicMetadataReaderIT.scala:70)
```

**Comment by morazow** [comment]
Hey @millin,

Thanks for the changes! I am going to check it soon.

File: src/it/scala/com/exasol/cloudetl/kafka/docker/KafkaImportIT.scala [hunk 3]
```diff
@@ -35,6 +40,25 @@ class KafkaImportIT extends BaseKafkaDockerIntegrationTest with BeforeAndAfterEa
     deleteTopic(topicName)
   }
 
+  test("import longs as timestamp values") {
+    case class TimestampRecord(timestamp: Long)
+    implicit val timestampRecordValueSerde = valueAvroSerde[TimestampRecord](getExternalSchemaRegistryUrl())
+
+    TimeZone.setDefault(exasolContainer.getTimeZone())
```
Comment (morazow) [inline src/it/scala/com/exasol/cloudetl/kafka/docker/KafkaImportIT.scala:47]: :+1:

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/FieldConverter.scala [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package com.exasol.cloudetl.kafka.deserialization
+
+import java.lang.{Long => JLong}
+import java.sql.Timestamp
+
+class FieldConverter(outputColumnTypes: Seq[Class[_]]) {
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/FieldConverter.scala:6]: Could you please add short javadoc?

**Review by morazow — CHANGES_REQUESTED** [review]
Hello @millin,

I left some minor suggestions. Please pull the branch before adding changes.

The tests were fine. I think you issue may be because of conflicting ports, for failure of docker container, or timing issue. You could also check if there is enough disk space for containers. I could not reproduce it. Let us observe it, and we can fix if it occurs regularly.

**Comment by millin** [comment]
Hello @morazow,

Thanks for the fix, code looks better :+1: 
Let me know if I need to fix anything else.

**Review by morazow — APPROVED** [review]
